subscription_oca/models/sale_subscription.py

```diff
@@ -0,0 +1,834 @@
+
+    @staticmethod
+    def string2datetime(string_date):
+        return datetime.strptime(string_date, ODOO_DATE_FORMAT)
+
+    def cron_subscription_management(self):
+        today = date.today()
+        query = []
+        for subscription in self.search(query):
+            if subscription.in_progress:
+                if (
+                    subscription.recurring_next_date == today
+                    and subscription.sale_subscription_line_ids
+                ):
+                    try:
+                        subscription.generate_invoice()
+                        subscription._onchange_template_id()
+                    except Exception:
+                        logger.exception("Error during Subscriptions Management cron")
+                if not subscription.recurring_rule_boundary:
+                    if subscription.date == today:
+                        subscription.action_close_subscription()
+
+            else:
+                if subscription.date_start == today:
+                    subscription.action_start_subscription()
+                    subscription.generate_invoice()
+                    subscription._onchange_template_id()
+
+    @api.depends("sale_subscription_line_ids")
+    def _compute_total(self):
+        for record in self:
+            recurring_total = amount_tax = 0.0
+            for order_line in record.sale_subscription_line_ids:
+                recurring_total += order_line.price_subtotal
+                amount_tax += order_line.amount_tax_line_amount
+            record.update(
+                {
+                    "recurring_total": recurring_total,
+                    "amount_tax": amount_tax,
+                    "amount_total": recurring_total + amount_tax,
+                }
+            )
+
+    def _compute_name(self):
+        template_code = self.template_id.code if self.template_id.code else ""
+        code = "{}".format(self.code) if self.code else ""
+        slash = "/" if template_code and code else ""
+        self.name = "{}{}{}".format(template_code, slash, code)
+
+    @api.onchange("template_id", "date_start")
+    def _onchange_template_id(self):
+        self._compute_name()
+        today = date.today()
+        if self.date_start:
+            today = self.date_start
+        if self.template_id.recurring_rule_boundary == "unlimited":
+            self.date = False
+            self.recurring_rule_boundary = True
+        else:
+            self.date = (
+                relativedelta(months=+self.template_id.recurring_rule_count)
+                + self.date_start
+            )
+            self.recurring_rule_boundary = False
+        self.terms = self.template_id.description
+        if self.template_id and self.account_invoice_ids_count > 0:
+            self.calculate_recurring_next_date(self.recurring_next_date)
+        else:
+            self.calculate_recurring_next_date(today)
+
+    def calculate_recurring_next_date(self, start_date):
```

Comment:
@carlos-domatix For these kind of computations, have you looked at https://github.com/OCA/server-ux/tree/13.0/base_recurrence ?

That one could be useful to lower the amount of code here

Reply:
@rousseldenis any chance to migrate it to 15? :)

Reply:
@francesco-ooops Feel free to do it. It should be quite easy. But no time to do it now.

Reply:
@carlos-domatix FYI we've migrated base_recurrence [here](https://github.com/OCA/server-ux/pull/701), you can add a dependency to that PR to employ it here

Reply:
Hi @carlos-domatix , just to say the migration is complete :)